pkg/controller/handlers/slackreceiver/slackreceiver.go

```diff
@@ -47,3 +71,146 @@ func CreateOAuthApp(req router.Request, _ router.Response) error {
 
 	return nil
 }
+
+func (s *Handler) SubscribeToSlackEvents(req router.Request, resp router.Response) error {
+	slackReceiver := req.Object.(*v1.SlackReceiver)
+
+	if s.subscribed[slackReceiver.Name] != nil {
+		return nil
+	}
+
+	cred, err := s.gptScript.RevealCredential(req.Ctx, []string{system.OAuthAppPrefix + slackReceiver.Spec.ThreadName}, string(types.OAuthAppTypeSlack))
+	if err != nil {
+		return err
+	}
+
+	slackAppToken := cred.Env["APP_TOKEN"]
+
+	if slackAppToken == "" {
+		return nil
+	}
+
+	slackClient := slack.New("", slack.OptionAppLevelToken(slackAppToken))
+	client := socketmode.New(slackClient)
+
+	h := &eventHandler{
+		ctx:     req.Ctx,
+		storage: s.storage,
+	}
+
+	socketmodeHandler := socketmode.NewSocketmodeHandler(client)
+
+	socketmodeHandler.Handle(socketmode.EventTypeConnecting, middlewareConnecting)
+	socketmodeHandler.Handle(socketmode.EventTypeConnectionError, middlewareConnectionError)
+	socketmodeHandler.Handle(socketmode.EventTypeConnected, middlewareConnected)
+
+	// Handle a specific event from EventsAPI
+	socketmodeHandler.HandleEvents(slackevents.AppMention, h.middlewareAppMentionEvent)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	s.subscribed[slackReceiver.Name] = cancel
```

Comment:
This needs a lock to protect against concurrent reads and writes.

Reply:
added lock to protect read/write